lib/generators/logidze/install/functions/logidze_logger.sql

```diff
@@ -70,7 +70,7 @@ CREATE OR REPLACE FUNCTION logidze_logger() RETURNS TRIGGER AS $body$
         END IF;
       END IF;
 
-      IF NEW = OLD THEN
```

Comment:
It would be great to see the performance difference of introducing this conversion; maybe, we can run our benchmarks? (See the `bench/` folder)

We can apply a kind of deoptimization here: 

```sql
BEGIN
  IF NEW = OLD THEN
    return NEW;
  END IF;
EXCEPTION
  WHEN <whatever it is> THEN
     IF to_jsonb(NEW.*) = to_jsonb(OLD.*) THEN
       RETURN NEW;
     END IF';
END;
````

(And we should also check if there is a noticeable performance degradation)

Reply:
So, I've updated the benchmarks, and the results seem to be similar for all of the options (when running `benchmarks/update_bench.rb`).

Original implementation (with JSONB `dump` column instead of JSON in order to avoid the error):
```
Generating 1000 records... Done in 31.6806233s
Warming up --------------------------------------
     Plain UPDATE #1    11.000  i/100ms
        PT UPDATE #1     8.000  i/100ms
   Logidze UPDATE #1    11.000  i/100ms
Calculating -------------------------------------
     Plain UPDATE #1    192.704  (±29.1%) i/s -    902.000  in   5.006673s
        PT UPDATE #1    119.735  (±33.4%) i/s -    560.000  in   5.047500s
   Logidze UPDATE #1    161.817  (±24.7%) i/s -    770.000  in   5.041253s

Comparison:
     Plain UPDATE #1:      192.7 i/s
   Logidze UPDATE #1:      161.8 i/s - same-ish: difference falls within error
        PT UPDATE #1:      119.7 i/s - same-ish: difference falls within error

Warming up --------------------------------------
     Plain UPDATE #2     6.000  i/100ms
        PT UPDATE #2     4.000  i/100ms
   Logidze UPDATE #2     5.000  i/100ms
Calculating -------------------------------------
     Plain UPDATE #2     83.782  (±27.5%) i/s -    402.000  in   5.065633s
        PT UPDATE #2     53.830  (±26.0%) i/s -    260.000  in   5.046071s
   Logidze UPDATE #2     80.952  (±48.2%) i/s -    370.000  in   5.047402s

Comparison:
     Plain UPDATE #2:       83.8 i/s
   Logidze UPDATE #2:       81.0 i/s - same-ish: difference falls within error
        PT UPDATE #2:       53.8 i/s - same-ish: difference falls within error
```

With casting to JSONB:
```
Generating 1000 records... Done in 32.0015556s
Warming up --------------------------------------
     Plain UPDATE #1    11.000  i/100ms
        PT UPDATE #1     8.000  i/100ms
   Logidze UPDATE #1    10.000  i/100ms
Calculating -------------------------------------
     Plain UPDATE #1    178.512  (±29.7%) i/s -    825.000  in   5.003882s
        PT UPDATE #1    123.008  (±26.8%) i/s -    584.000  in   5.047852s
   Logidze UPDATE #1    155.165  (±25.8%) i/s -    740.000  in   5.055674s

Comparison:
     Plain UPDATE #1:      178.5 i/s
   Logidze UPDATE #1:      155.2 i/s - same-ish: difference falls within error
        PT UPDATE #1:      123.0 i/s - same-ish: difference falls within error

Warming up --------------------------------------
     Plain UPDATE #2     5.000  i/100ms
        PT UPDATE #2     4.000  i/100ms
   Logidze UPDATE #2     5.000  i/100ms
Calculating -------------------------------------
     Plain UPDATE #2     89.237  (±51.5%) i/s -    400.000  in   5.046479s
        PT UPDATE #2     53.057  (±28.3%) i/s -    252.000  in   5.004583s
   Logidze UPDATE #2     77.411  (±28.4%) i/s -    365.000  in   5.001165s

Comparison:
     Plain UPDATE #2:       89.2 i/s
   Logidze UPDATE #2:       77.4 i/s - same-ish: difference falls within error
        PT UPDATE #2:       53.1 i/s - same-ish: difference falls within error
```

With casting to JSONB only in case of an error:
```
Generating 1000 records... Done in 31.6082378s
Warming up --------------------------------------
     Plain UPDATE #1    10.000  i/100ms
        PT UPDATE #1     8.000  i/100ms
   Logidze UPDATE #1    10.000  i/100ms
Calculating -------------------------------------
     Plain UPDATE #1    181.856  (±31.3%) i/s -    840.000  in   5.026085s
        PT UPDATE #1    123.663  (±27.5%) i/s -    584.000  in   5.040383s
   Logidze UPDATE #1    159.170  (±22.0%) i/s -    760.000  in   5.004458s

Comparison:
     Plain UPDATE #1:      181.9 i/s
   Logidze UPDATE #1:      159.2 i/s - same-ish: difference falls within error
        PT UPDATE #1:      123.7 i/s - same-ish: difference falls within error

Warming up --------------------------------------
     Plain UPDATE #2     5.000  i/100ms
        PT UPDATE #2     4.000  i/100ms
   Logidze UPDATE #2     5.000  i/100ms
Calculating -------------------------------------
     Plain UPDATE #2     82.446  (±27.9%) i/s -    390.000  in   5.006902s
        PT UPDATE #2     54.852  (±23.7%) i/s -    264.000  in   5.050243s
   Logidze UPDATE #2     74.415  (±24.2%) i/s -    355.000  in   5.002545s

Comparison:
     Plain UPDATE #2:       82.4 i/s
   Logidze UPDATE #2:       74.4 i/s - same-ish: difference falls within error
        PT UPDATE #2:       54.9 i/s - same-ish: difference falls within error
```

Here's the SQL code I used for the last option (for reference):
```sql
BEGIN
	IF NEW = OLD THEN
		RETURN NEW;
	END IF;
EXCEPTION
	WHEN undefined_function THEN
		IF to_jsonb(NEW.*) = to_jsonb(OLD.*) THEN
			RETURN NEW;
		END IF;
END;
```

Reply:
> Let's bump the version for this SQL

@baygeldin ping

Reply:
Sorry, this PR completely slipped my mind. I've updated the version in the beginning of the file (what's its purpose btw?). I can also revert changes to benchmarks that were needed to compare different implementations if it's needed.

Reply:
Oh, I see, thought it was just a comment initially.